fix: :bug: fix error where llm rerank failed (relevant document not found)

upchunk opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # ( fix error where llm rerank failed (relevant document not found))

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [x] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x] No

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

logan-markewich commented on llama-index-core/llama_index/core/indices/utils.py:
```diff
@@ -90,24 +92,31 @@ def default_parse_choice_select_answer_fn(
+    except Exception as e:
+        _logger.error(f"Error in LLM Rerank: {e!s}")
+        for i in answer_lines:
+            _logger.error(i)
```
does this actually fix the error? won't this return an empty array now? 

imo it was probably safer for end users to be raising an error 🤔

upchunk replied:
Its to handle when something like this happened

![image](https://github.com/user-attachments/assets/41f2712a-8f6c-41db-901d-1fe6ca1da1e0)

upchunk replied:
It is a rare case but it's still brok my app. i think, returning what the LLM really says will really help the user to mitigae what kind of issue they're facing, instead of just "list index is out of range"

logan-markewich replied:
Yes but now we are hiding the error in a log statement, and then returning zero nodes. A user might think everything worked fine if they have logs turned off.

Better to raise an error, and the user can handle try/except. We can make the error better by doing something like 

`raise ValueError("Unable to parse LLM output: {output})`

upchunk replied:
Nice Idea, I'll update the code for sure. Thanks 👍

upchunk commented:
Nice idea 💡, i'll update the cose

Pada Jum, 16 Agu 2024, 22.17, Logan ***@***.***> menulis:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In llama-index-core/llama_index/core/indices/utils.py
> <https://github.com/run-llama/llama_index/pull/15399#discussion_r1719985310>
> :
>
> > -        if answer_num > num_choices:
> -            continue
> -        answer_nums.append(answer_num)
> -        # extract just the first digits after the colon.
> -        _answer_relevance = re.findall(r"\d+", line_tokens[1].split(":")[1].strip())[0]
> -        answer_relevances.append(float(_answer_relevance))
> +            answer_nums.append(answer_num)
> +            # extract just the first digits after the colon.
> +            _answer_relevance = re.findall(
> +                r"\d+", line_tokens[1].split(":")[1].strip()
> +            )[0]
> +            answer_relevances.append(float(_answer_relevance))
> +    except Exception as e:
> +        _logger.error(f"Error in LLM Rerank: {e!s}")
> +        for i in answer_lines:
> +            _logger.error(i)
>
> Yes but now we are hiding the error in a log statement, and then returning
> zero nodes. A user might think everything worked fine if they have logs
> turned off.
>
> Better to raise an error, and the user can handle try/except. We can make
> the error better by doing something like
>
> raise ValueError("Unable to parse LLM output: {output})
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/run-llama/llama_index/pull/15399#discussion_r1719985310>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOZ2YM4A2454U6LJUWEHOFTZRYJY7AVCNFSM6AAAAABMR5C4E2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDENBSHE2TINRVGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

logan-markewich commented:
@upchunk I think my suggestion still stands, but this PR is very old now, so going to close this out. Feel free to re-open with the changes!
